Uchu.World/Objects/Components/ReplicaComponents/BuffComponent.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using RakDotNet.IO;
+
+namespace Uchu.World
+{
+    public class BuffInfo
+    {
+        public uint BuffId;
+        public uint DurationSecs;
+        public bool CancelOnUnequip;
+        public bool CancelOnRemoveBuff;
+        public bool CancelOnUi;
+        public bool CancelOnZone;
+        public bool CancelOnDamaged;
+        public bool CancelOnDeath;
+        public bool CancelOnLogout;
+        public bool ApplyOnTeammates;
+        public bool UseRefCount;
+    }
+
+    public class BuffComponent : ReplicaComponent
+    {
+        public override ComponentId Id => ComponentId.BuffComponent;
+
+        private readonly List<BuffInfo> _buffs = new List<BuffInfo>();
+
+        protected BuffComponent()
+        {
+
+        }
+
+        public void AddBuff(BuffInfo buffInfo)
+        {
+            _buffs.Add(buffInfo);
+
+            if (!GameObject.TryGetComponent<DestroyableComponent>(out var destroyable)) return;
+            if (buffInfo.CancelOnDeath)
+            {
+                Delegate onDeathDelegate = null;
+                onDeathDelegate = Listen(destroyable.OnDeath, () =>
+                {
+                    ReleaseListener(onDeathDelegate);
+                    RemoveBuff(buffInfo);
+                });
+            }
+
+            if (buffInfo.CancelOnDamaged)
+            {
+                // this looks ugly, maybe there's a better way to listen once
+                Delegate cancelOnDamaged1 = null;
+                Delegate cancelOnDamaged2 = null;
+                cancelOnDamaged1 = Listen(destroyable.OnArmorChanged, (newValue, delta) =>
```

Comment:
Can't really think of anything than changing the api of Listen to include a `once = false` parameter

Reply:
We may need to do a bit more for this, like create a listener group that allows storing a set of listeners that can be disconnected at once. For this case, we listen to 2 listeners and disconnect both when received.